### ISSUE TITLE
Add dynamic styles to Choropleths?

### ISSUE THREAD

**Issue by deeplook** [jupyter-widgets/ipyleaflet#454]
At the moment a Choropleth accepts only one kind of data (in the `choro_data` parameter) to control the fill color of GeoJSON features. I figured it would be helpful to extend this further and control the entire styling of single features. I my case I'm most interested in setting line weights, but setting other style attributes would be also useful.

I've looked at [Leaflet's Choropleth example](https://leafletjs.com/examples/choropleth/) and tweaked it only a bit to generate the effect I would like to see in the image below.

<img width="667" alt="Screenshot 2020-01-04 at 12 41 04" src="https://user-images.githubusercontent.com/1001778/71765118-e0658600-2ef0-11ea-98a0-575d5f741325.png">

This was done by modifying the `style` function in a minimal way (the `weight` attribute):

```javascript
	function style(feature) {
		return {
			// weight: 2,
			weight: Math.random() * 10, 
			opacity: 1,
			color: 'white',
			dashArray: '3',
			fillOpacity: 0.7,
			fillColor: getColor(feature.properties.density)
		};
	}
```

Ideally one could implement this with a new dynamic `style` parameter replacing the existing `choro_data`, `colormap`, and `border_color`, which would break existing code. So maybe add such a more generic `style` parameter and respect (and deprecate) the previous parameters for some time before removing them? The new `style` parameter could be a mapping from `id` keys to style dicts and/or a function returning such style dicts.

I think this would add quite a few steroids to the existing Choropleths, and I'm glad to help with the implmentation. Opinions?

**Comment by martinRenou** [comment]
This could be done easily actually. The Choropleth layer in ipyleaflet is just a GeoJSON layer, that we modify (fully in Python) before sending to the front-end: https://github.com/jupyter-widgets/ipyleaflet/blob/master/ipyleaflet/leaflet.py#L603. So it's "only" Python code, which should be easier to modify for newcomers.

**Comment by omanges** [comment]
I would like to work on this, @martinRenou  please assign this to me.

**Comment by martinRenou** [comment]
Woaw, thanks a lot!! Looking forward to this :)

**Comment by omanges** [comment]
I think here user can just pass an entire dict of styles with key as the feature ID, which can be set to the style attribute of each Feature at this step https://github.com/jupyter-widgets/ipyleaflet/blob/9066d48edf31f917519b8c486f05061367c34890/ipyleaflet/leaflet.py#L618 correct me if I am wrong  @martinRenou :)

**Comment by martinRenou** [comment]
Sure! Maybe here a python callback could be used? Like the user would pass a function looking like this:
```python
import random
def compute_style(feature, colormap):
    return {
        'fillColor': colormap(feature),
        'weight': random.randint(1, 5),
    }
```

Also, the feature id `"id"` is hardcoded. It would be better to make it a trait of the Choropleth class.
Something like:

```python
class Choropleth(...):

    key = Unicode()

    ... later in the code
    feature['properties']['style'] = dict(fillColor=color_dict[feature[self.key]],
```
But this can be fixed separately.

**Comment by omanges** [comment]
Hi, @martinRenou , think there is already a trait called style in one of its superclasses so should I create a new parameter for Choropleth like `style_callback`, where the user can pass a callback function as you specified.

**Comment by martinRenou** [comment]
Fixed by #518
